chainercv/links/model/faster_rcnn/faster_rcnn.py

```diff
@@ -0,0 +1,321 @@
+# Copyright (c) 2016 Shunta Saito
+# Licensed under The MIT License [see LICENSE for details]
+# https://github.com/mitmul/chainer-faster-rcnn
+# --------------------------------------------------------
+# Faster R-CNN
+# Copyright (c) 2015 Microsoft
+# Licensed under The MIT License [see LICENSE for details]
+# Written by Ross Girshick and Sean Bell
+# https://github.com/rbgirshick/py-faster-rcnn
+# --------------------------------------------------------
+
+from __future__ import division
+
+import numpy as np
+
+import chainer
+from chainer import cuda
+import chainer.functions as F
+from chainercv.links.model.faster_rcnn.utils.loc2bbox import loc2bbox
+from chainercv.utils import non_maximum_suppression
+
+from chainercv.transforms.image.resize import resize
+
+
+class FasterRCNN(chainer.Chain):
+
+    """Base class for Faster R-CNN.
+
+    This is a base class for Faster R-CNN links supporting object detection
+    API [#]_. The following three stages constitute Faster R-CNN.
+
+    1. **Feature extraction**: Images are taken and their \
+        feature maps are calculated.
+    2. **Region Proposal Networks**: Given the feature maps calculated in \
+        the previous stage, produce set of RoIs around objects.
+    3. **Localization and Classification Heads**: Using features that belong \
+        to the proposed RoIs, classify the categories of the objects in the \
+        RoIs and improve localizations.
+
+    Each stage is carried out by one of the callable
+    :class:`chainer.Chain` objects :obj:`feature`, :obj:`rpn` and :obj:`head`.
+
+    There are two functions :func:`predict` and :func:`__call__` to conduct
+    object detection.
+    :func:`predict` takes images and returns bounding boxes that are converted
+    to image coordinates. This will be useful for a scenario when
+    Faster R-CNN is treated as a black box function, for instance.
+    :func:`__call__` is provided for a scnerario when intermediate outputs
+    are needed, for instance, for training and debugging.
+
+    Links that support obejct detection API have method :func:`predict` with
+    the same interface. Please refer to :func:`FasterRCNN.predict` for
+    further details.
+
+    .. [#] Shaoqing Ren, Kaiming He, Ross Girshick, Jian Sun. \
+    Faster R-CNN: Towards Real-Time Object Detection with \
+    Region Proposal Networks. NIPS 2015.
+
+    Args:
+        extractor (callable Chain): A callable that takes a BCHW image
+            array and returns a feature.
+        rpn (callable Chain): A callable that has the same interface as
+            :class:`chainercv.links.RegionProposalNetwork`. Please refer to
+            the documentation found there.
+        head (callable Chain): A callable that takes
+            a BCHW array, RoIs and batch indices for RoIs. This returns class
+            dependent localization paramters and class scores.
+        n_fg_class (int): The number of classes excluding the background.
```

Comment:
We can get the number of classes from `head`. This argument looks redundant and may cause mismatch issue.